File: queues/servicebus/PHP/lib/ServiceBusRest.php
<?php
declare(strict_types=1);
/*
 * Copyright Core DF — Apache License 2.0
 * Azure Service Bus REST (SharedAccessSignature).
 */
final class ServiceBusRest
{
    private static function parseConnString(string $conn): array
    {
        $parts = [];
        foreach (explode(';', $conn) as $piece) {
            [$k, $v] = array_pad(explode('=', $piece, 2), 2, '');
            $parts[$k] = $v;
        }
        $endpoint = str_replace('sb://', 'https://', $parts['Endpoint'] ?? '');
        return [
            'endpoint' => rtrim($endpoint, '/'),
            'key_name' => $parts['SharedAccessKeyName'] ?? '',
            'key' => $parts['SharedAccessKey'] ?? '',
        ];
    }

    private static function sasToken(string $resourceUri, string $keyName, string $key): string
    {
        $expiry = (string)(time() + 3600);
        $encoded = rawurlencode($resourceUri);
        $stringToSign = $encoded . "\n" . $expiry;
        $signature = base64_encode(hash_hmac('sha256', $stringToSign, $key, true));
        return 'SharedAccessSignature sr=' . $encoded
            . '&sig=' . rawurlencode($signature)
            . '&se=' . $expiry
            . '&skn=' . rawurlencode($keyName);
    }

    public static function sendMessage(string $conn, string $queue, mixed $body): array
    {
        $cfg = self::parseConnString($conn);
        $resource = $cfg['endpoint'] . '/' . $queue;
        $url = $resource . '/messages';
        $payload = is_string($body) ? $body : json_encode($body);
        $ch = curl_init($url);
        curl_setopt_array($ch, [
            CURLOPT_POST => true,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER => [
                'Authorization: ' . self::sasToken($resource, $cfg['key_name'], $cfg['key']),
                'Content-Type: application/json',
            ],
            CURLOPT_POSTFIELDS => $payload,
        ]);
        $resp = curl_exec($ch);
        $code = (int)curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        if ($code >= 400) {
            return ['status_code' => $code, 'error' => $resp ?: 'send failed'];
        }
        return ['status_code' => 200];
    }

    public static function receiveMessages(string $conn, string $queue, float $timeoutSec, int $maxMessages, bool $complete): array
    {
        $cfg = self::parseConnString($conn);
        $resource = $cfg['endpoint'] . '/' . $queue;
        $messages = [];
        for ($i = 0; $i < max(1, $maxMessages); $i++) {
            $url = $resource . '/messages/head?timeout=' . (int)$timeoutSec;
            $ch = curl_init($url);
            curl_setopt_array($ch, [
                CURLOPT_POST => true,
                CURLOPT_RETURNTRANSFER => true,
                CURLOPT_HEADER => true,
                CURLOPT_HTTPHEADER => [
                    'Authorization: ' . self::sasToken($resource, $cfg['key_name'], $cfg['key']),
                ],
            ]);
            $raw = curl_exec($ch);
            $code = (int)curl_getinfo($ch, CURLINFO_HTTP_CODE);
            curl_close($ch);
            if ($code === 204 || $raw === false || $raw === '') {
                break;
            }
            if ($code >= 400) {
                return ['status_code' => $code, 'error' => $raw];
            }
            [$headers, $body] = explode("\r\n\r\n", $raw, 2);
            $lock = null;
            if (preg_match('/BrokerProperties:\s*(.+)/i', $headers, $m)) {
                $props = json_decode(trim($m[1]), true);
                $lock = $props['LockToken'] ?? null;
            }
            $value = json_decode($body, true);
            if ($value === null) {
                $value = $body;
            }
            $messages[] = ['queue' => $queue, 'value' => $value];
            if ($complete && $lock) {
                $delUrl = $resource . '/messages/' . $lock;
                $del = curl_init($delUrl);
                curl_setopt_array($del, [
                    CURLOPT_CUSTOMREQUEST => 'DELETE',
                    CURLOPT_RETURNTRANSFER => true,
                    CURLOPT_HTTPHEADER => [
                        'Authorization: ' . self::sasToken($resource, $cfg['key_name'], $cfg['key']),
                    ],
                ]);
                curl_exec($del);
                curl_close($del);
            }
        }
        return ['status_code' => 200, 'messages' => $messages];
    }
}

File: queues/servicebus/Ruby/lib/service_bus_rest.rb
# Copyright Core DF — Apache License 2.0
# Azure Service Bus REST helpers (SharedAccessSignature).

require 'json'
require 'net/http'
require 'uri'
require 'cgi'
require 'base64'
require 'openssl'

module ServiceBusRest
  module_function

  def parse_conn_string(conn)
    parts = conn.split(';').map { |p| p.split('=', 2) }.to_h
    endpoint = parts['Endpoint'].to_s.sub(%r{^sb://}, 'https://').delete_suffix('/')
    {
      endpoint: endpoint,
      key_name: parts['SharedAccessKeyName'],
      key: parts['SharedAccessKey'],
    }
  end

  def sas_token(resource_uri, key_name, key)
    expiry = (Time.now.to_i + 3600).to_s
    encoded = CGI.escape(resource_uri)
    string_to_sign = "#{encoded}\n#{expiry}"
    signature = Base64.strict_encode64(OpenSSL::HMAC.digest('sha256', key, string_to_sign))
    "SharedAccessSignature sr=#{encoded}&sig=#{CGI.escape(signature)}&se=#{expiry}&skn=#{CGI.escape(key_name)}"
  end

  def send_message(conn, queue, body)
    cfg = parse_conn_string(conn)
    resource = "#{cfg[:endpoint]}/#{queue}"
    url = URI("#{resource}/messages")
    req = Net::HTTP::Post.new(url)
    req['Authorization'] = sas_token(resource, cfg[:key_name], cfg[:key])
    req['Content-Type'] = 'application/json'
    req.body = body.is_a?(String) ? body : JSON.generate(body)
    resp = Net::HTTP.start(url.host, url.port, use_ssl: true) { |h| h.request(req) }
    return { status_code: resp.code.to_i, error: resp.body } if resp.code.to_i >= 400
    { status_code: 200 }
  end

  def receive_messages(conn, queue, timeout_sec, max_messages, complete)
    cfg = parse_conn_string(conn)
    resource = "#{cfg[:endpoint]}/#{queue}"
    messages = []
    max_messages.times do
      url = URI("#{resource}/messages/head?timeout=#{timeout_sec.to_i}")
      req = Net::HTTP::Post.new(url)
      req['Authorization'] = sas_token(resource, cfg[:key_name], cfg[:key])
      resp = Net::HTTP.start(url.host, url.port, use_ssl: true) { |h| h.request(req) }
      break if resp.code.to_i == 204 || resp.body.to_s.empty?
      return { status_code: resp.code.to_i, error: resp.body } if resp.code.to_i >= 400
      lock = resp['BrokerProperties'] && JSON.parse(resp['BrokerProperties'])['LockToken'] rescue nil
      value = begin; JSON.parse(resp.body); rescue; resp.body; end
      messages << { queue: queue, message_id: resp['BrokerProperties'], value: value }
      if complete && lock
        del = URI("#{resource}/messages/#{lock}")
        del_req = Net::HTTP::Delete.new(del)
        del_req['Authorization'] = sas_token(resource, cfg[:key_name], cfg[:key])
        Net::HTTP.start(del.host, del.port, use_ssl: true) { |h| h.request(del_req) }
      end
    end
    { status_code: 200, messages: messages }
  end
end
